obspy/clients/fdsn/mass_downloader/utils.py

```diff
@@ -0,0 +1,534 @@
+                                                                str(e)))
+
+
+def get_stationxml_contents(filename):
+    """
+    Really fast way to get all channels with a response in a StationXML
+    file. Sometimes it is too expensive to parse the full file with ObsPy.
+    This is usually orders of magnitudes faster.
+
+    Will only returns channels that contain response information.
+
+    :param filename: The path to the file.
+    :returns: list of ChannelAvailability objects.
+    """
+    # Small state machine.
+    network, station, location, channel, starttime, endtime = [None] * 6
+
+    ns = "http://www.fdsn.org/xml/station/1"
+    network_tag = "{%s}Network" % ns
+    station_tag = "{%s}Station" % ns
+    channel_tag = "{%s}Channel" % ns
+    response_tag = "{%s}Response" % ns
+
+    context = etree.iterparse(filename, events=("start", ),
+                              tag=(network_tag, station_tag, channel_tag,
+                                   response_tag))
+
+    channels = []
+    for event, elem in context:
+        if elem.tag == channel_tag:
+            channel = elem.get('code')
+            location = elem.get('locationCode').strip()
+            starttime = obspy.UTCDateTime(elem.get('startDate'))
+            end_date = elem.get('endDate')
+            if end_date is not None:
+                endtime = obspy.UTCDateTime(elem.get('endDate'))
+            else:
+                endtime = obspy.UTCDateTime(2599, 1, 1)
+        elif elem.tag == response_tag:
+            channels.append(ChannelAvailability(
+                network, station, location, channel, starttime, endtime,
+                filename))
+        elif elem.tag == station_tag:
+            station = elem.get('code')
+            location, channel, starttime, endtime = \
+                None, None, None, None
+        elif elem.tag == network_tag:
+            network = elem.get('code')
+            station, location, channel, starttime, endtime = \
+                None, None, None, None, None
+        while elem.getprevious() is not None:
+            del elem.getparent()[0]
+
+    return channels
+
+
+def get_stationxml_filename(str_or_fct, network, station, channels,
+                            starttime, endtime):
+    """
+    Helper function getting the filename of a StationXML file.
+
+    :param str_or_fct: The string or function to be evaluated.
+    :type str_or_fct: function or str
+    :param network: The network code.
+    :type network: str
+    :param station: The station code.
+    :type station: str
+    :param channels: The channels. Each channel is a tuple of two strings:
+        location code and channel code.
+    :type channels: list of tuples
+    :param starttime: The start time.
+    :type starttime: :class:`~obspy.core.utcdatetime.UTCDateTime`
+    :param endtime: The end time.
+    :type endtime: :class:`~obspy.core.utcdatetime.UTCDateTime`
+
+    The rules are simple, if it is a function, network, station, channels,
```

Comment:
This clause appears to continue on after "function", even though that's the end of the condition, so something like "if it is a function, then the network, station, ..." would be better.